include/seqan3/alignment/matrix/debug_matrix.hpp

```diff
@@ -0,0 +1,370 @@
+        return col_width;
+    }
+
+protected:
+
+    //!\brief Whether the entry_type is trace_directions.
+    static constexpr bool is_traceback_matrix = std::is_same_v<std::decay_t<entry_type>, trace_directions>;
+
+    //!\brief Same as at(*row*, *col*), but as string.
+    std::string entry_at(size_t const row, size_t const col, fmtflags2 flags) const noexcept
+    {
+        format_type const & symbols = (flags & fmtflags2::utf8) == fmtflags2::utf8 ? unicode : csv;
+
+        entry_type const & entry = at(row, col);
+        if (!is_traceback_matrix && entry == matrix_inf<entry_type>)
+            return symbols.inf;
+        return as_string(entry, flags);
+    }
+
+    //!\brief Convert a value into a std::string
+    template <typename entry_type>
+    static std::string as_string(entry_type && entry, fmtflags2 const flags) noexcept
+    {
+        std::stringstream strstream;
+        debug_stream_type stream{strstream};
+        stream << flags << entry;
+        return strstream.str();
+    }
+
+    //!\brief The length of the *str* (traceback symbols are unicode aware)
+    //!\sa https://en.wikipedia.org/wiki/UTF-8 for encoding details
+    static size_t unicode_str_length(std::string const & str) noexcept
+    {
+        size_t length = 0u;
+        for (auto it = str.cbegin(), it_end = str.cend(); it < it_end; ++it, ++length)
+        {
+            uint8_t v = *it;
+            if ((v & 0b11100000) == 0b11000000)
+                ++it;
+            else if ((v & 0b11110000) == 0b11100000)
+                it += 2;
+            else if ((v & 0b11111000) == 0b11110000)
+                it += 3;
+        }
+        return length;
+    }
+
+    //!\brief Format used by seqan3::detail::debug_matrix.
+    struct format_type
+    {
+        //!\brief epsilon symbol (a single symbol)
+        char const * epsilon{};
+        //!\brief column separator symbol (a single symbol)
+        char const * col_sep{};
+        //!\brief row separator symbol (a single symbol)
+        char const * row_sep{};
+        //!\brief row separator symbol (a single symbol)
+        char const * row_col_sep{};
+        //!\brief infinity symbol (a single symbol)
+        char const * inf{};
+    };
+
+    //!\brief The format when printing to a ascii stream.
+    static constexpr format_type csv{" ", ";", "", "", ""};
+    //!\brief The format when printing to a unicode stream.
+    static constexpr format_type unicode{"ε", "║", "═", "╬", "∞"};
+
+public:
+    //!\brief What is the width (number of chars) of an entry. Defaults to #auto_column_width.
+    std::optional<size_t> column_width{std::nullopt};
+
+protected:
+    //!\brief The matrix
+    matrix_t _matrix;
+    //!\brief The first sequence of the sequence alignment.
+    sequence1_t _sequence1;
+    //!\brief The second sequence of the sequence alignment.
+    sequence2_t _sequence2;
+
+    //!\brief Befriend the test case
+    friend struct debug_matrix_print_test;
```

Comment:
now the library befriends a google test :open_mouth: we don't want test stuff in our library. 
And if this is detail code it does not need to (should not?) be tested anyway?

Reply:
Everything should be tested...

Reply:
I removed it; I designed the class as protected so I can inherit it in test cases to access the protected members.